Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp

```diff
@@ -395,82 +397,117 @@ String WebsiteDataStore::migrateMediaKeysStorageIfNecessary(const String& direct
+
 #if ENABLE(ARKIT_INLINE_PREVIEW)
-    if (!configuration->modelElementCacheDirectory().isEmpty())
-        resolvedConfiguration->setModelElementCacheDirectory(resolveAndCreateReadWriteDirectoryForSandboxExtension(configuration->modelElementCacheDirectory()));
+    if (!directories.modelElementCacheDirectory.isEmpty())
+        directories.modelElementCacheDirectory = resolveAndCreateReadWriteDirectoryForSandboxExtension(directories.modelElementCacheDirectory);
 #endif
-    if (!configuration->searchFieldHistoryDirectory().isEmpty())
-        resolvedConfiguration->setSearchFieldHistoryDirectory(resolveAndCreateReadWriteDirectoryForSandboxExtension(configuration->searchFieldHistoryDirectory()));
 
-    // Resolve file paths.
-    if (!configuration->cookieStorageFile().isEmpty()) {
-        auto resolvedCookieDirectory = resolveAndCreateReadWriteDirectoryForSandboxExtension(FileSystem::parentPath(configuration->cookieStorageFile()));
-        resolvedConfiguration->setCookieStorageFile(FileSystem::pathByAppendingComponent(resolvedCookieDirectory, FileSystem::pathFileName(configuration->cookieStorageFile())));
+    if (!directories.cookieStorageFile.isEmpty()) {
+        auto resolvedCookieDirectory = resolveAndCreateReadWriteDirectoryForSandboxExtension(FileSystem::parentPath(directories.cookieStorageFile));
+        directories.cookieStorageFile = FileSystem::pathByAppendingComponent(resolvedCookieDirectory, FileSystem::pathFileName(directories.cookieStorageFile));
     }
+}
 
-    // Default paths of WebsiteDataStore created with identifier are not under caches or tmp directory,
-    // so we need to explicitly exclude them from backup.
-    if (configuration->identifier()) {
-        Vector<String> allCacheDirectories = {
-            resolvedMediaCacheDirectory()
-            , resolvedNetworkCacheDirectory()
+WebsiteDataStoreConfiguration::Directories WebsiteDataStore::resolvedDirectories()
+{
+    Locker resolveLocker { m_resolveLock };
+    auto result = m_resolvedDirectories;
+    RELEASE_ASSERT(result);
+    return *result;
+}
+
+void WebsiteDataStore::dispatchResolveDirectories()
+{
+    ASSERT(RunLoop::isMain());
+
+    m_hasDispatchedResolveDirectories = true;
+    Ref resolveDirectoriesQueue = WorkQueue::create("com.apple.WebKit.WebsiteDataStore.resolveDirectories", WorkQueue::QOS::UserInteractive);
+    resolveDirectoriesQueue->dispatch([this, protectedThis = Ref { *this }, directories = m_configuration->directories().isolatedCopy()]() mutable {
+        resolveDirectories(directories);
+        {
+            Locker resolveLocker { m_resolveLock };
+            m_resolvedDirectories = crossThreadCopy(directories);
+            m_resolveCondition.notifyOne();
+        }
+        RunLoop::main().dispatch([protectedThis = WTFMove(protectedThis)] {
+            protectedThis->didResolveDirectories();
+        });
+    });
+}
+
+void WebsiteDataStore::didResolveDirectories()
```

Comment:
Could you explain why this must be called on the main thread? This wasn't obvious to me.
Ideally, we'd call this straight from the resolve queue.

Reply:
Because we are dispatching the tasks to a work queue with lower priority. The resolve directories queue has the same priority as main thread to avoid blocking main thread, but these tasks don't need that.

Reply:
oh you are asking why we cannot do the task dispatch on the dispatch work queue, not why we do the task dispatch. Yes I think that's possible; will change.

Reply:
Correct. You can dispatch to your lower priority queue straight from the high priority queue, without involving the main thread.